File: eratosthenes/eratosthenes0.cc
#include "eratosthenes0.h"

#include <cmath>

void Eratosthenes0::generate(int64 x) {
  flags_.clear();
  if (x > 100000000)
    return;

  flags_.resize(x, true);

  flags_[0] = false;
  flags_[1] = false;
  const int64 sqrt_x = std::ceil(std::sqrt(x) + 0.1);
  for (int64 p = 2; p < sqrt_x; ++p) {
    if (!flags_[p])
      continue;
    for (int64 mult = p * p; mult < x; mult += p)
      flags_[mult] = false;
  }
}

int64 Eratosthenes0::count() {
  if (flags_.empty())
    return -1;

  int64 ret = 0;
  for (bool f : flags_) {
    if (f)
      ++ret;
  }
  return ret;
}

File: eratosthenes/rust/src/eratosthenes/eratosthenes0.rs
use crate::bits::Bits;

use super::Eratosthenes;

pub struct Eratosthenes0 {
    flags: Bits,
}

impl Eratosthenes0 {
    pub fn new() -> Eratosthenes0 {
        Eratosthenes0 {
            flags: Bits::new(0, false),
        }
    }
}

impl Eratosthenes for Eratosthenes0 {
    fn version(&self) -> i32 {
        0
    }

    fn generate(&mut self, x: usize) {
        let mut flags = Bits::new(x + 1, true);
        flags.unset(0);
        flags.unset(1);
        let sqr_x = (x as f64).sqrt().ceil() as usize;
        for p in 2..=sqr_x {
            if flags.get(p) {
                for j in ((p * p)..=x).step_by(p) {
                    flags.unset(j);
                }
            }
        }
        self.flags = flags;
    }

    fn generate_range(&mut self, _x: usize, _y: usize) {
        eprintln!("Does not support ranged generation.");
    }

    fn count(&self) -> usize {
        self.flags.bit_count()
    }
}
